torch/csrc/jit/passes/constant_propagation.cpp

```diff
@@ -75,39 +75,35 @@ struct ConstantPropagator {
     }
   }
 
-  std::vector<IValue> runNode(Node* n) {
+  c10::optional<std::vector<IValue>> runNode(Node* n) {
     auto op = n->getOperation();
     Stack stack;
     for (auto input : n->inputs()) {
       stack.push_back(*toIValue(input));
     }
-    op(stack);
-    auto var_outputs = fmap(stack, [&](IValue v) -> IValue {
+    try {
+      op(stack);
+    } catch (...) {
+      return c10::nullopt;
+    }
+    for (const IValue& v : stack) {
       if (v.isTensor()) {
-        auto t = std::move(v).toTensor();
-        if (t.defined()) {
-          if (t.requires_grad()) {
-            // error gets caught within propagateNode()
-            throw c10::Error("Can't insert requires grad as constant", "");
-          }
-          return IValue(t);
-        } else {
-          return t;
+        at::Tensor t  = v.toTensor();
+        if (t.defined() && t.requires_grad()) {
+          // requires grad tensors cannot be constants
+          return c10::nullopt;
         }
-      } else {
-        return v;
       }
-    });
-    return var_outputs;
+    }
+    return stack;
   }
 
   void propagateNode(Node* n) {
     std::vector<IValue> outputs;
-    try {
-      outputs = runNode(n);
-    } catch (...) {
-      // Catch exceptions. This op may not be run,
-      // so catch the error here & leave the op in the graph
+    if (auto outputs_opt = runNode(n)) {
```

Comment:
Nice to be rid of try/catch, and I'm assuming it's because Constant was the only op that might throw? But I'm not sure what motivated the optionalizing of the return value.

Reply:
I am localizing the try/catch logic inside of runNode since it is an implementation detail of how we do constant prop by actually run the potentially wrong operator. I moved the getOperator call outside of the try/catch because that was actually a bug: getOperator should always succeed but we were silently letting it fail. It matters in the next PR where I introduce a bunch of things where getOperator would fail.